util/query_xivapi.ts

```diff
@@ -0,0 +1,551 @@
+
+              Example:  ID > 30, Patch != 88, Name ~ (Topaz Titan, Ruby Carb*)
+ `,
+  });
+
+  queryParser.addArgument(['-e', '--endpoint'], {
+    nargs: 1,
+    type: 'string',
+    help: 'The name of the XIVAPI endpoint to query',
+  });
+
+  queryParser.addArgument(['-c', '--columns'], {
+    nargs: 1,
+    type: 'string',
+    help: 'Columns/parameters to obtain from the endpoint',
+  });
+
+  queryParser.addArgument(['-f', '--filter'], {
+    nargs: '+',
+    type: 'string',
+    help: 'Filter(s) to apply on returned data',
+  });
+
+  queryParser.addArgument(['-ll', '--loglevel'], {
+    nargs: 1,
+    type: 'string',
+    choices: logLevels.map((ll) => ll[0]),
+    help: `The level of console output you want to see (default: ${_LOGLEVEL_DEFAULT})`,
+  });
+};
+
+const queryApi = async (
+  endpoint: string,
+  columns: string,
+  filters: string,
+  loglevel: LogLevelKey,
+): Promise<void> => {
+  const outputToConsole = (result: XivApiQueryResult): void => {
+    log.printNoHeader(JSON.stringify(result, null, 2).replace(/"/g, '\''));
+  };
+
+  log.setLogLevel(loglevel);
+
+  const _ENDPOINT = endpoint.trim();
+  if (endpoint.length < 1) {
+    log.fatalError('No endpoint specified.');
+    return;
+  }
+
+  const _COLUMNS = columns.replace(/\s+/g, '').split(',').filter((c) => c !== '');
+  if (_COLUMNS.length < 1 || _COLUMNS[0] === undefined) {
+    log.fatalError('No columns specified.');
+    return;
+  }
+
+  const api = new XivApi(null, log);
+
+  const apiData = await api.queryApi(
+    _ENDPOINT,
+    _COLUMNS,
+  ) as XivApiQueryResult;
+
+  if (filters === '') {
+    log.debug('No filters were specified.');
+    if (apiData.length === 0)
+      log.printNoHeader('No API data matched your query.');
+    else
+      outputToConsole(apiData);
+  } else {
+    log.debug('Applying specified filters to data...');
+    const filteredData = applyFilters(apiData, parseFilters(filters));
+    if (filteredData.length === 0)
+      log.printNoHeader('No API data matched your query & filter conditions.');
+    else
+      outputToConsole(filteredData);
+  }
+};
+
+const parseFilters = (filters: string): FilterSet => {
+  const parsedFilters: FilterSet = {
+    includes: [],
+    excludes: [],
+  };
+
+  // filters are comma-separated, but don't count commas inside (); that's an array.
+  const regexFilterSeparator = /(?<!\([\w\s]+),(?![\w\s]+\))/;
+  const rawFilters = filters.split(regexFilterSeparator).filter((f) => f !== '');
+
+  const regexColumn = /^\s*(\w+)\s*/;
+  const regexOp = new RegExp(`(${allOps.join('|')})`);
+  const regexValueArr = /\s*[([][\w\s*]+,[\w\s*]+[,\w\s*]*[)\]]\s*/;
+  const regexValueStr = /[\w\s*]+/;
+  const regexValueEither = new RegExp(`(${regexValueStr.source}$|${regexValueArr.source}$)`);
+  const regexFilterParts = new RegExp(
+    `${regexColumn.source}${regexOp.source}${regexValueEither.source}`,
+  );
+
+  rawFilters.forEach((rf) => {
```

Comment:
There's a lot of effort here to parse out filter information. Would it make more sense to just have `--filters` accept a JSON object, for example something like these:
```json
[{"column":"ID","operator":">","value":30}]
[{"column":"ID","operator":">","value":30},{"column":"Name","operator":"~","value":["Topaz Titan","Ruby Carb*"]}]
{"column":"ID","operator":">","value":30}
```

(These could also be "JSON-like", as was done for timeline netregex parsing)

Mostly I suggest this because regex is a pain to debug later down the line whenever it gets even slightly complicated.

Reply:
> There's a lot of effort here to parse out filter information. Would it make more sense to just have `--filters` accept a JSON object, for example something like these:

I've noodled on this for a bit, but I think I lean toward keeping as-is.  Changing to JSON-style input would remove the initial capture regex (although some will still be needed to handle wildcards).  But most of the code in that `parseFilters` will still be needed to validate user input, regardless of whether it comes as a string or JSON.  If anything, the regex saves some work by rejecting most non-conforming input.

I guess I see higher value in the simplicity of short text for quick CLI input, rather than typing out (what could be rather lengthy) JSON.  If there were a text editor involved, I think it'd be easier, but a CLI makes a lengthy string hard to read, check for errors, and edit after a certain length.  Particularly if there's multiple filters, that's a lot of characters devoted to repeating 'column', 'operator', 'value' over and over.

Reply:
I think I lean mildly toward "keep it as it is and do not change it to accept JSON-likes" myself. I agree with valarnin that maintenance on regex-based parsing can be a real pain, but as a user of the utility, it's a lot easier to use if I don't have to worry about formatting a bunch of JSON on the command line. Particularly since we're looking at 7.0 in 6-8 months, I think it's worth taking on a little more burden for maintainers in order to save a little time later when *using* the script in the new-content time crunch.

Reply:
Yeah, I've given this some more thought over the past couple days and I think it's fine to keep it as is.

Maybe we should encourage adding comments with specific examples of "match" and "don't match" for any regex above a certain complexity going forward?

Reply:
> Maybe we should encourage adding comments with specific examples of "match" and "don't match" for any regex above a certain complexity going forward?

Yup, makes sense to me.  I think the filter separator and the value-array regex are both fugly enough to warrant comments, so I'll add some before merging.  I think the others are probably simple enough not to need comments, but maybe an overall comment on the structure of a proper filter would be useful.